rasterio/rio/calc.py

```diff
@@ -0,0 +1,148 @@
+# Calc command.
+
+import logging
+import math
+import os.path
+import re
+import sys
+import warnings
+
+import click
+import snuggs
+
+import rasterio
+from rasterio.fill import fillnodata
+from rasterio.features import sieve
+from rasterio.rio.cli import cli
+
+
+def get_bands(inputs, d, i=None):
+    """Get a rasterio.Band object from calc's inputs"""
+    path = inputs[d] if d in dict(inputs) else inputs[int(d)-1][1]
+    if i:
+        return rasterio.band(rasterio.open(path), i)
+    else:
+        src = rasterio.open(path)
+        return [rasterio.band(src, i) for i in src.indexes]
+
+
+def read_array(ix, subix=None, dtype=None):
+    """Change the type of a read array"""
+    arr = snuggs._ctx.lookup(ix, subix)
```

Comment:
@sgillies 2 comments here:

- looks like you are using a internal API from the signature ```._ctx```; if you are using it outside ```snuggs```, it probably should be a public API, e.g., ```snuggs.ctx.lookup()``` or similar.  (going here w/ the convention that leading underscores denote internal functions / objects)
- some extra validation would be nice here, with simpler error messages.  I can make this produce a long exception using:

```$ rio calc "(read 2)" tests/data/RGB.byte.tif /tmp/test.tif```

```read: invalid band index``` would be a fine message in this case.